src/lib/blkdata_svc/blk_read_tracker.cpp

```diff
@@ -98,8 +94,16 @@ void BlkReadTracker::merge(const BlkId& blkid, int64_t new_ref_count,
 void BlkReadTracker::insert(const BlkId& blkid) { merge(blkid, 1, nullptr); }
 void BlkReadTracker::remove(const BlkId& blkid) { merge(blkid, -1, nullptr); }
 
-void BlkReadTracker::wait_on(const BlkId& blkid, after_remove_cb_t&& after_remove_cb) {
-    merge(blkid, 0, std::make_shared< blk_track_waiter >(std::move(after_remove_cb)));
+void BlkReadTracker::wait_on(MultiBlkId const& blkids, after_remove_cb_t&& after_remove_cb) {
+    if (blkids.num_pieces() == 1) {
+        merge(blkids, 0, std::make_shared< blk_track_waiter >(std::move(after_remove_cb)));
+    } else {
+        auto waiter = std::make_shared< blk_track_waiter >(std::move(after_remove_cb));
```

Comment:
There will be race, right? When a read finishes on this specifc blkid but not yet for other blkids, the remove_cb will be called immaturely by freeing all the blkids in vdev?

Reply:
We are having a single waiter instance for all `blkids`, so if any read waits for a specific blkid, they will only call waiter after all refcount goes to zero.

In other words, isn't this same scenario as in existing implementation. If a blkid is bigger than what it can hold, it already splits into multiple smaller blkids and adding a reference. Now that is extended to multiple pieces of BlkIds right?

Reply:
Say we 2 separate blks addressed by this MultiBlkId. one blkid is merged and waiter waits on it. Before this thread got chance to merge the 2nd (line: 104), read completes on the 1st one and waiter got zero ref_cnt and waiter will be gone and trigger the callback immaturely. 

Existing implementation is a one-blk-id, one waiter pattern, here we have multiple blk-ids, one waiter, which breaks the assumption blkreader track is working. 

Now this assumption is not there and one pending blk read completes could affect (free) all the other blks in vdev.

Reply:
We will call the callback only on the destructor of `blk_track_waiter` right? The scenario you mentioned will not happen before all the blk_ids are added, because `blk_track_waiter` will still be in scope.

Reply:
every merge operation will push the waiter into the BlkReadTrack record, and once one merge operation is done, there is nothing can prevent read completes routine to call m_read_blk_tracker->remove(blkid, xxx) which will trigger a deref to that record, in return destruct the waiter.

Reply:
I take back the race, we are holding the shared_ptr of waiter. 
But we are forcing all blk ids to be freed as a whole for each single blk to be able to be freed in vdev, which is un-necessary, right? In fact, each blk-id should be able to get freed once its pending blkid finished their reads.